File: ruby/interpreter.rb
class Interpreter
  def initialize(rpn)
    @rpn = rpn
    @stack = []
  end

  def run
    @rpn.each do |item|
      if item[0] == :NUMBER
        @stack << item[1]
      elsif item[0] == :OPERATOR
        operator = item[1]
        number1 = @stack.pop
        number2 = @stack.pop
        @stack << apply(operator, number1, number2)
      end
    end

    @stack.first
  end

  private

  def apply(operator, number1, number2)
    case operator
    when "*"
      number1 * number2
    when "/"
      number1 / number2
    when "-"
      number1 - number2
    when "+"
      number1 + number2
    end
  end
end

File: go/interpreter/interpreter.go
package interpreter

import (
	"../parser"
	"math"
	"strconv"
)

type Interpreter struct {
	Rpn   []parser.Node
	Stack []float64
}

func popNumberFromStack(interpreter *Interpreter) float64 {
	length := len(interpreter.Stack)
	number := interpreter.Stack[length-1]
	interpreter.Stack = interpreter.Stack[:length-1]
	return number
}

func NewInterpreter(nodes []parser.Node) Interpreter {
	return Interpreter{
		Rpn:   nodes,
		Stack: make([]float64, 0),
	}
}

func (interpreter *Interpreter) Run() float64 {
	for _, parsedToken := range interpreter.Rpn {
		switch parsedToken.(type) {
		case parser.Number:
			parsedNumber, _ := strconv.ParseFloat(parsedToken.NodeValue(), 64)
			interpreter.Stack = append(interpreter.Stack, parsedNumber)
		case parser.Operator:
			operator := parsedToken.NodeValue()
			number2 := popNumberFromStack(interpreter)
			number1 := popNumberFromStack(interpreter)
			result := apply(operator, number1, number2)
			interpreter.Stack = append(interpreter.Stack, result)
		}
	}

	result := interpreter.Stack[0]
	return result
}

func apply(operator string, number1 float64, number2 float64) float64 {
	switch operator {
	case "/":
		return number1 / number2
	case "*":
		return number1 * number2
	case "+":
		return number1 + number2
	case "-":
		return number1 - number2
	case "^":
		return math.Pow(number1, number2)
	}
	// fallback that should not be hit
	return 0
}
